### ISSUE TITLE
kstest, ks_2samp: confusing `mode` argument descriptions

### ISSUE THREAD

**Issue by brandondavid** [scipy/scipy#10963]
For the `mode` parameter, [kstest](https://docs.scipy.org/doc/scipy-1.3.1/reference/generated/scipy.stats.kstest.html) has:  
`'approx' : use approximation to exact distribution of test statistic`

Whereas [ks_2samp](https://docs.scipy.org/doc/scipy-1.3.1/reference/generated/scipy.stats.ks_2samp.html) has:
`'exact' : use approximation to exact distribution of test statistic`

- It seems odd that the antonyms _'approx'_ and _'exact'_ would have the same description.
- When `mode='approx'`, [kstest](https://github.com/scipy/scipy/blob/v1.3.1/scipy/stats/stats.py#L4817-L4822) uses an asymptotic or exact distribution directly.
- When `mode='exact'`, [ks_2samp](https://github.com/scipy/scipy/blob/v1.3.1/scipy/stats/stats.py#L5518-L5536) attempts to make an exact computation.

See https://hackmd.io/3Gj_8GsYTuCa-xGBD9KHog for detailed notes and initial discussion.

_On the recommendation of @rgommers, creating this issue and pinging @pvanmulbregt to ask for clarification._

**Comment by josef-pkt** [comment]
In case there is anything still based on my old changes

```
if mode == 'approx':
    pval_two = distributions.kstwobign.sf(D * np.sqrt(N))
    if N > 2666 or pval_two > 0.80 - N*0.3/1000:
        return KstestResult(D, pval_two)
    else:
        return KstestResult(D, 2 * distributions.ksone.sf(D, N))
```

AFAIR
When I rewrote this originally, then the justification was mostly "it works", it was an approximation to the exact distribution from R or an R package that I could get with the special functions that were available in scipy at the time. The thresholds were heuristic based on several examples, but I don't remember any details. Here it was mainly "we get approximately the same results as exact in R" without theoretical justification.

In other or newer cases we might actually get the "exact" distribution or an approximation that is (theoretically) justified as being a close approximation of the exact distribution.

**Comment by josef-pkt** [comment]
looks like the above snippet is my code
https://github.com/scipy/scipy/commit/3e3e8eaef5ff26148c3f79c925b4d20b15e05988#diff-cc307f50f4eb0f9917b55fdf1dbb9b8cR2063

I cannot find the old scipy list right now to see whether I wrote any justification
https://github.com/scipy/scipy/issues/922#issuecomment-17020445

**Comment by josef-pkt** [comment]
one detail reading around code and hackmd

 stats.ksone is exact (correct in small samples) for 1-sided test D+ or D- but not for the two-sided test.
The `approx` is that we use twice the exact 1-sided p-value, i.e. ksone, as approximation to the two-sided small sample p-value.
`return KstestResult(D, 2 * distributions.ksone.sf(D, N))`

Aside: the asymptotic distribution is approximately exact in large samples.

**Comment by brandondavid** [comment]
Thank you for all the insight, @josef-pkt!

Please correct me if I'm wrong, but my understanding is that `ksone` is indeed the exact distribution.  If so, then I don't understand the phrase "approximation to exact distribution".  If not, then the docs for [special.smirnov](https://docs.scipy.org/doc/scipy-1.3.1/reference/generated/scipy.special.smirnov.html) are misleading.  I'm a bit of out of my depth here, but it seems to me that the only approximation to be had with kstest() is when it uses the asymptotic distribution, which naturally results in approximate values (with the degree of approximation depending on the sample size).

**Comment by brandondavid** [comment]
Whoops, I should've refreshed before posting.  I think you anticipated and answered my question.

I think some of these clarifications should be included in the docstring, so I will make some edits for review.  Thanks!

**Comment by brandondavid** [comment]
### kstest
It would be good to have the behavior of _'approx'_ documented in the "Notes" section of the docstring, but for the "Parameters" of [kstest](https://github.com/scipy/scipy/blob/v1.3.1/scipy/stats/stats.py#L4705-L4709), I propose:
```
    * 'approx': approximate the exact distribution of test statistic
    * 'asymp': use asymptotic distribution of test statistic
```
This is just a personal preference, as the phrase "approximation to exact" could be confusing for certain users.


### ks_2samp
ks_2samp and its helper functions implement much of ["The Significance Probability of the Smirnov Two-Sample Test"](https://projecteuclid.org/download/pdf_1/euclid.afm/1485893310) by J.L. Hodges.  Here is some pseudocode for an abbreviated version of ks_2samp, with reference to Hodges' paper and terminology:
```
function ks_2samp:
    if 'exact' mode:
        if 'two-sided' alternative:
            if "square":
                use Hodges 2.4 "reflectional method" for P2 (closed expression)
            else:
                use Hodges 2.1 "inside method" for P2 (direct computation)
        else:
            if "square":
                use Hodges 2.3 "reflectional method" for P1 (closed expression)
            else:
                use Hodges 2.2 "outside method" for P1 (direct computation)
    if 'asymp' mode:
        if 'two-sided' alternative:
            use stats.kstwobign (asymptotic distribution)
        else:
            use Hodges 5.3 "interpolation formula" (from 4.10 "asymptotic expression")
```

For _'exact'_, I think "use direct computation" (which would include closed expressions) is more accurate than "use approximation to exact distribution" (which implies something like `ksone` is being called).  And given the number of checks where ks_2samp fails to _'asymp'_ mode, perhaps "attempt direct computation" would be more helpful.

For _'asymp'_, I leave it to someone else to decide whether ks_2samp truly uses the asymptotic distribution for one-sided tests.  By my reading of Hodges, the 5.3 "interpolation formula" follows from 4.10, which is an "asymptotic expression" developed from the same "reflectional method" used to produce the closed expressions 2.3 and 2.4.  But given that 4.10 follows from 4.5, which has the size of one sample going to infinity, I assume that's equivalent enough to Kolmogorov's limiting form that the _'asymp'_ description should remain unchanged.

For _'auto'_, I find "small samples" to be more helpful than "small size arrays" given the context.

So for [ks_2samp](https://github.com/scipy/scipy/blob/v1.3.1/scipy/stats/stats.py#L5402-L5408), I propose the following, with all other implementation details discussed in the "Notes" section of the docstring:
```
    * 'auto' : use 'exact' for small samples, 'asymp' for large samples
    * 'exact' : attempt direct computation of test statistic
    * 'asymp' : use asymptotic distribution of test statistic
```

**Comment by pvanmulbregt** [comment]
I agree the doc is confusing.  I have a pending PR, dependent on gh-9719, which uses `kstwo`, `kstwobign` and `ksone` in the expected manner and also cleans up some of the differences in documentation of `ks_2samp` and `kstest` to match.  I'll compare that with your suggestions above.
